### PR TITLE
Feature net zero electricity 2030 eag

### PR DESCRIPTION
**Pull request by pworschischek-aggmag**
closes https://github.com/AGGM-AG/pypsa-at-planning/issues/40 

Before asking for a review for this PR make sure to complete the following checklist:

## Changes proposed in this Pull Request
Adds the net-zero electricity production constraint.

## Checklist

**Required:**
- [x] Changes are tested locally and behave as expected.
- [x] Code and workflow changes are documented.
- [x] A brief description of the changes has been added to `Changelog.AT.md`.
- [x] All Sourcery Bot review suggestions have been implemented or discarded with an explanation.
- [x] All github actions succeed.

**If applicable:**
- [ ] Changes in configuration options are reflected in `scripts/lib/validation`.
- [ ] For new data sources or versions, [these instructions](https://pypsa-eur.readthedocs.io/en/latest/data_sources.html) have been followed.
- [ ] New rules are documented in the appropriate `docs-at/` files.

## Summary by Sourcery

Introduce national net-zero electricity and green fuel production constraints for PyPSA-AT, wire them into the DE additional functionality, and add configuration, statistics support, and tests to enforce Austria’s 2030 net-zero electricity balance.

New Features:
- Add orchestrated PyPSA-AT constraint application including national CO2 budgets, solar utility trajectories, and net-zero electricity constraints.
- Implement national net-zero electricity production constraints with companion green gas, hydrogen, and methanol production constraints based on fuel and bus carriers.
- Provide configuration options for enabling net-zero electricity constraints and specifying start years and admissible green fuel sources per country.
- Add documentation pages describing regulatory net-zero electricity requirements and their implementation in PyPSA-AT.

Bug Fixes:
- Ensure national CO2 budget tests respect configuration toggles and fail early when budgets are disabled or inconsistent.
- Fix gas pipeline extendability logic to be controlled by configuration and applied consistently for both existing and new pipelines, and always stop Ukrainian gas transit in the prenetwork step.

Enhancements:
- Refactor PyPSA statistics configuration and custom groupers into file I/O utilities so they are registered when networks are read.
- Add helper utilities and fixtures for accessing configuration from solved networks and for constructing synthetic networks used in constraint unit tests.
- Extend evaluation utilities to support input-share calculations without output scaling, enabling fuel-attributable demand statistics.
- Integrate PyPSA-AT constraints via a single entry point from the PyPSA-DE additional functionality script.

Documentation:
- Document Austrian regulatory requirements for net-zero electricity, including legal context, constraint design, configuration, and limitations, and add a regulatory requirements section to the documentation navigation.

Tests:
- Add extensive unit and statistics-based tests for net-zero electricity and green fuel constraints, including contract tests for gas bus usage and verification via pypsa.statistics.
- Add fixtures for loading result networks and building a minimal synthetic EAG-compliant network for constraint testing.
- Add regression tests ensuring loads never act as energy suppliers and that renewable-fuel bus topologies remain constant across myopic networks.

### CODE REVIEW

**Comment by pworschischek-aggmag** [comment]
@sourcery-ai review

File: scripts/pypsa-de/additional_functionality.py [hunk 1]
```diff
@@ -889,18 +884,8 @@ def additional_functionality(n, snapshots, snakemake):
...
-    else:
-        logger.warning("No national CO2 budget specified!")
-
     if investment_year == 2020:
         adapt_nuclear_output(n)
 
-    if snakemake.config["mods"]["PEMMDB_trajectories"].get("enable"):
-        add_solar_utility_trajectory_constraints(n, snakemake, investment_year)
+    # PyPSA-AT constraints
+    pypsa_at_constraints(n, snakemake, investment_year)
```
Comment (SourceryAI) [inline scripts/pypsa-de/additional_functionality.py:891]: **question (bug_risk):** Calling `pypsa_at_constraints` unconditionally can break for configs without the new keys.

This function now assumes keys like `snakemake.params.solving["constraints"]["co2_budget_national"]` and `snakemake.config["mods"]["net_zero_electricity"]` always exist with a specific structure. For older or minimal configs, this will raise `KeyError`s. If you need to keep those configs working, add guards (feature flags or key existence checks) here or inside `pypsa_at_constraints` so it degrades gracefully when those sections are missing.

File: docs-at/explanations/regulatory-requirements/eag-net-zero-electricity.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+**Left-hand side** — domestic green H2 production:
+
+Links where `name.startswith(country)`, `bus1` ∈ Austrian H2 buses, and `bus0` carrier is in
+the configurable `mods.net_zero_electricity.h2_sources` allowlist. The default allowlist is
+`{"AC", "low voltage", "solid biomass"}`, admitting H2 Electrolysis and biomass-to-H2 as
+green producers.
+
+Why an allowlist and not a "`bus0` ≠ H2" mirror of the gas/methanol pattern: SMR / SMR CC
+(`bus0 = gas`, `bus1 = H2`) is blue/grey H2. A "`bus0` ≠ H2" filter would silently admit
+fossil-gas H2 onto the green-H2 LHS. The allowlist makes the policy decision explicit.
```
Comment (SourceryAI) [inline docs-at/explanations/regulatory-requirements/eag-net-zero-electricity.md:200]: **issue (typo):** Use a plural verb for the compound subject “SMR / SMR CC”.

Since “SMR / SMR CC” refers to two processes, update “is blue/grey H2” to “are blue/grey H2” for grammatical agreement.

```suggestion
Why an allowlist and not a "`bus0` ≠ H2" mirror of the gas/methanol pattern: SMR / SMR CC
(`bus0 = gas`, `bus1 = H2`) are blue/grey H2. A "`bus0` ≠ H2" filter would silently admit
fossil-gas H2 onto the green-H2 LHS. The allowlist makes the policy decision explicit.
```

✅ Addressed in 9d024f9489b9dd4ca195b76a5072d3b1e0943526: The documentation now reads “SMR / SMR CC (`bus0 = gas`, `bus1 = H2`) are blue/grey H2,” correcting the plural verb agreement.

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 2 issues, and left some high level feedback:

- In mods/__init__.py you replaced the exported name 'override_nuts' with 'pypsa_at_constraints'; if any downstream code or configs still rely on 'override_nuts', consider re‑exporting both to avoid a breaking change in the public mods API.
- add_national_net_zero_electricity_constraints assumes config["mods"]["net_zero_electricity"] is always present; if you want this feature to be optional, it would be safer to use .get with a default and short‑circuit cleanly when the section is absent rather than raising a KeyError.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- In mods/__init__.py you replaced the exported name 'override_nuts' with 'pypsa_at_constraints'; if any downstream code or configs still rely on 'override_nuts', consider re‑exporting both to avoid a breaking change in the public mods API.
- add_national_net_zero_electricity_constraints assumes config["mods"]["net_zero_electricity"] is always present; if you want this feature to be optional, it would be safer to use .get with a default and short‑circuit cleanly when the section is absent rather than raising a KeyError.

## Individual Comments

### Comment 1
<location path="scripts/pypsa-de/additional_functionality.py" line_range="890-891" />
<code_context>

-    if snakemake.config["mods"]["PEMMDB_trajectories"].get("enable"):
-        add_solar_utility_trajectory_constraints(n, snakemake, investment_year)
+    # PyPSA-AT constraints
+    pypsa_at_constraints(n, snakemake, investment_year)
</code_context>
<issue_to_address>
**question (bug_risk):** Calling `pypsa_at_constraints` unconditionally can break for configs without the new keys.

This function now assumes keys like `snakemake.params.solving["constraints"]["co2_budget_national"]` and `snakemake.config["mods"]["net_zero_electricity"]` always exist with a specific structure. For older or minimal configs, this will raise `KeyError`s. If you need to keep those configs working, add guards (feature flags or key existence checks) here or inside `pypsa_at_constraints` so it degrades gracefully when those sections are missing.
</issue_to_address>

### Comment 2
<location path="docs-at/explanations/regulatory-requirements/eag-net-zero-electricity.md" line_range="197-199" />
<code_context>
+`{"AC", "low voltage", "solid biomass"}`, admitting H2 Electrolysis and biomass-to-H2 as
+green producers.
+
+Why an allowlist and not a "`bus0` ≠ H2" mirror of the gas/methanol pattern: SMR / SMR CC
+(`bus0 = gas`, `bus1 = H2`) is blue/grey H2. A "`bus0` ≠ H2" filter would silently admit
+fossil-gas H2 onto the green-H2 LHS. The allowlist makes the policy decision explicit.
+
</code_context>
<issue_to_address>
**issue (typo):** Use a plural verb for the compound subject “SMR / SMR CC”.

Since “SMR / SMR CC” refers to two processes, update “is blue/grey H2” to “are blue/grey H2” for grammatical agreement.

```suggestion
Why an allowlist and not a "`bus0` ≠ H2" mirror of the gas/methanol pattern: SMR / SMR CC
(`bus0 = gas`, `bus1 = H2`) are blue/grey H2. A "`bus0` ≠ H2" filter would silently admit
fossil-gas H2 onto the green-H2 LHS. The allowlist makes the policy decision explicit.
```
</issue_to_address>
~~~

</details>

***

Hi @pworschischek-aggmag! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

File: scripts/pypsa-de/additional_functionality.py (hunk 1 above)
Comment (pworschischek-aggmag) [reply scripts/pypsa-de/additional_functionality.py:891]: This is the orchestrator. It shall always be called. individual constraints are skipped at per-function level.

File: docs-at/explanations/regulatory-requirements/eag-net-zero-electricity.md (hunk 2 above)
Comment (pworschischek-aggmag) [reply docs-at/explanations/regulatory-requirements/eag-net-zero-electricity.md:200]: done
